src/crafting_gui.cpp

```diff
@@ -202,9 +198,35 @@ const recipe *select_crafting_recipe( int &batch_size )
             if( batch ) {
                 batch_recipes( crafting_inv, helpers, current, available, chosen );
             } else {
-                // Set current to all recipes in the current tab; available are possible to make
-                pick_recipes( crafting_inv, helpers, current, available,
-                              tab.cur(), subtab.cur(), filterstring );
+
+                available.clear();
+                current.clear();
+
+                if( filterstring.empty() ) {
+                    current = recipe_dict.in_category( tab.cur(), subtab.cur() != "CSC_ALL" ? subtab.cur() : "" );
+                } else {
+                    current = recipe_dict.search( filterstring );
+                }
+
+                // cache recipe availability on first display
+                for( const auto e : current ) {
+                    if( !availability_cache.count( e ) ) {
+                        availability_cache.emplace( e, e->can_make_with_inventory( crafting_inv, helpers ) );
+                    }
+                }
+
+                std::stable_sort( current.begin(), current.end(), []( const recipe * a, const recipe * b ) {
+                    return b->difficulty < a->difficulty;
+                } );
+
+                std::stable_sort( current.begin(), current.end(), [&]( const recipe * a, const recipe * b ) {
+                    return availability_cache[a] && !availability_cache[b];
+                } );
+
+                std::transform( current.begin(), current.end(),
+                std::back_inserter( available ), [&]( const recipe * e ) {
+                    return e->can_make_with_inventory( crafting_inv, helpers );
```

Comment:
> @codemime any suggestions as to this PR as I'm stuck for other obvious performance improvements (sans spaghetti-code)

You create the `availability_cache`, but, as far as I can say, never really use it (apart from sorting the `current` list). This line recalculates everything that was cached.

I'm gonna take a closer look at the algorithm. For now I'd suggest per-tab caching (to eliminate any recalculation after each tab was cached). Cache and display the first tab, meanwhile (while waiting for the input) cache the rest in the background.